nav2_costmap_2d/test/unit/keepout_filter_test.cpp

```diff
@@ -0,0 +1,363 @@
+// Copyright (c) 2020 Samsung Research Russia
```

Comment:
I still fail to see what this is testing. `InfoPublisher` doesn't do anything and exercises no code from the keepout filter. You do create a costmap in `TestNode` but I fail to see where any of that was used to validate the keepout filter. 

This doesn't need to be complicated. Create a layered costmap with a keepout filter. Publish the info message with an arbitrary topic name. On that topic, publish your mask with a blob of keepout costs in the middle and process to update it. Verify on your layered costmap that all those values are now non-empty. 

Then do the same for a rolling costmap with the published mask non-aligned with some overlap that it stops at the overlap.

Reply:
> I still fail to see what this is testing. InfoPublisher doesn't do anything and exercises no code from the keepout filter. You do create a costmap in TestNode but I fail to see where any of that was used to validate the keepout filter.

`InfoPublisher` is used only for publishing a `CostmapFilterInfo` message for a keepout filter, as well as `MaskPublisher` publishes a map mask for the filter. There is a `keepout_filter_` object of `nav2_costmap_2d::KeepoutFilter` type is placed in a `TestNode`. This is a unit test, so this object is manually created and initialized. In test body (`TestNode::testKeepoutFilter`), there is called main processing routine from keepout filter: `keepout_filter_->process()` and checking that it is processing given `mask_` correctly. Since it is unit test, it is not designed to check anything with rolling windows, bounds or something else in dynamics. For that system-test is made. This test is intended to verify that main functionality API from KeepoutFilter works correctly.

> This doesn't need to be complicated. Create a layered costmap with a keepout filter. Publish the info message with an arbitrary topic name. On that topic, publish your mask with a blob of keepout costs in the middle and process to update it. Verify on your layered costmap that all those values are now non-empty.

That is exactly that this test is making. During the `TestNode::createKeepoutFilter()` call, layered costmap is made:  `nav2_costmap_2d::LayeredCostmap layers(...)`. Then, keepout_filter is being initialized as a `Layer` with a reference to made `layers&`. Info message and map mask are being published in a `InfoPublisher` and `MaskPublisher`, as mentioned above. Then, after everyring is created, published, received and set, keepout filter is being called with a `process()` routine for different (overlapped/non-overlapped/out-of-map-mask) size windows `(min_i, min_j, max_i, max_j)`. After that, values are being checked that they are set correct and now non-empty in a `verifyMasterGrid()` routine.

> Then do the same for a rolling costmap with the published mask non-aligned with some overlap that it stops at the overlap.

Initially, mask and map are not aligned: map's origin is (0, 0) whether the mask's origin is (-5, -5).

Reply:
You don't really use the keepout mask, you use these points https://github.com/ros-planning/navigation2/blob/b36a5f3f74f1b799360e50e46715fa87b4310946/nav2_costmap_2d/test/unit/keepout_filter_test.cpp#L298 to check, but not the `mask_` you create presumably to publish to the filter https://github.com/ros-planning/navigation2/blob/b36a5f3f74f1b799360e50e46715fa87b4310946/nav2_costmap_2d/test/unit/keepout_filter_test.cpp#L207.

Reply:
In this test it is checking not whole overlapping of `mask_` with `master_grid_`, but only some intersection windows manually set in tests. Whole `mask_` area by design should never be passed to a `process()`, so it is impossible to verify the `master_grid_` by `mask_`. Therefore, `keepout_points_` were invented for result verification. It could be more, e.g the stack of marked bars which need to be checked at `master_grid_`, but in my opinion, it is much more complex for simple test.

Let's check the example for clarification: the `keepout_filter_->process(*master_grid_, 1, 3, 3, 6, pose);` called with `(min_i, min_j, max_i, max_j) = (1,3,3,6)` window will give only 4 points in `mask_` with `master_grid_` overlapping. These points only will be marked with `keepout_value` in `master_grid_` during the `process()` routine. Since I initially know the window size `(1,3,3,6)` manually set in the test, I also know the expected result after `process()` call will be finished: it will be 4 points belonging to chosen `(1,3,3,6)` area. Therefore, I think it is easy to manually add these 4 points to verification because them and only them should present on `master_grid_` after `process()` call.

Reply:
Attaching the picture drawn how the keepout_filter unit test works:
![P00923-162039](https://user-images.githubusercontent.com/60094858/94018318-23d28e00-fdb9-11ea-8d4e-4fc36c6b129b.jpg)
At this picture `master_grid_` is drawn by pencil, `mask_` is drawn by blue pen, window areas for `process(min_i, min_j, max_i, max_j)` calls are marked in dotted lined bars. Points which would finally appear in `master_grid_` with a `keepout_value` are marked as `X` by blue pen.
From this picture it is seen that test is checking all overlapped/on-edge/out-out-edge cases for `(min_i, min_j, max_i, max_j)` windows in order to get the best coverage of `process()` algorithms. Also, there are only 5 points to be changed at `master_grid_` during the `process()` calls. That is why, `verifyMasterGrid()` is decided to work with points instead of using something else.

Reply:
OK, I see what you were doing. What I thought was more "straight forward" was to publish this `mask_` with a known size / content, `process` the filter, and then check that all of the values in `mask_` are inputted into the master grid correctly. That way you don't have to selectively pick points and then there's no need to selectively check ambient points, but this works as well.